src/lib/components/SubscribeForm/index.js

```diff
@@ -0,0 +1,243 @@
+
+/**
+ * Element that renders newsletter subscription form.
+ *
+ * @extends {BaseElement}
+ * @final
+ */
+class SubscribeForm extends BaseElement {
+  static get properties() {
+    return {
+      errors: {
+        type: Object,
+      },
+      submitted: {
+        type: Boolean,
+      },
+    };
+  }
+
+  constructor() {
+    super();
+    this.checkboxes = ["WebDevNewsletter", "collects-pii-spii-checkbox"];
+    this.needsDoubleOptIn = [
+      "AT: Austria",
+      "DE: Germany",
+      "GR: Greece",
+      "LU: Luxembourg",
+      "NO: Norway",
+    ];
+    this.robotName = "is-it-just-me-or-was-this-form-filled-out-by-a-robot";
+    this.submissionUrl =
+      "https://services.google.com/fb/submissions/591768a1-61a6-4f16-8e3c-adf1661539da/";
+    this.submitted = false;
+  }
+
+  /**
+   * @param {FormData} form
+   * @return {FormData}
+   */
+  cleanForm(form) {
+    const country = form.get("Country");
+    this.checkboxes.forEach((box) => {
+      const optedIn = this.needsDoubleOptIn.includes(country)
+        ? "Unconfirmed"
+        : "True";
+      form.set(box, optedIn);
+    });
+
+    form.append("LanguagePreference", "en-US");
+    form.delete(this.robotName);
+
+    return form;
+  }
+
+  submit(form) {
+    return new Promise((resolve, reject) => {
+      const request = new XMLHttpRequest();
+      request.open("POST", this.submissionUrl);
+      request.send(form);
+      request.onload = function() {
+        const response = request.response ? JSON.parse(request.response) : {};
+
+        if (response && response.result === "accepted") {
+          resolve();
+        } else if (response && response.errors) {
+          reject(response.errors);
+        } else {
+          reject("Could not submit, please try again.");
+        }
+      };
+    });
+  }
+
+  handleSubmit(e) {
+    e.preventDefault();
+    if (this.submitted === false) {
+      const form = new FormData(e.target);
+      const formIsRobot = form.get(this.robotName).length !== 0;
+
+      if (!formIsRobot) {
+        const cleanedForm = this.cleanForm(form);
+        this.submit(cleanedForm)
+          .then(() => (this.submitted = true))
+          .catch((e) => (this.errors = e));
+      } else {
+        this.submitted = true;
+      }
+    }
+
+    return false;
+  }
+
+  renderForm() {
+    let errorMessage;
+
+    if (this.errors) {
+      if (typeof this.errors === "string") {
+        errorMessage = html`
+          <div class="web-subscribe-error">
+            <p>${this.errors}</p>
+          </div>
+        `;
+      } else {
+        const errorMessages = Object.values(this.errors).map(
+          (e) =>
+            html`
+              <p>${e}</p>
+            `,
+        );
+        errorMessage = html`
+          <div class="web-subscribe-error">
+            ${errorMessages}
+          </div>
+        `;
+      }
+    }
+
+    const countriesOptions = countries.map((country) => {
+      return country[1] === "United States"
```

Comment:
I'd love to use [geoIP headers](https://cloud.google.com/appengine/docs/flexible/dotnet/reference/request-headers) to choose the correct initial selection (I get that a user might change it).

Reply:
I don't think we can control that. It's a formbox endpoint, so I'm not sure what could be done here.

Reply:
We get access to the user's geolocation inside `server.js` as a header, we can do something with it there—perhaps pipe it back to this code via cookie, HTTP endpoint, ...